rust/worker/src/execution/operators/ranked_group_by.rs

```diff
@@ -0,0 +1,818 @@
+    fn extract_key(&self, keys: &[Key]) -> Vec<Option<MetadataValue>> {
+        keys.iter()
+            .map(|key| match key {
+                Key::MetadataField(field) => self.metadata.get(field).cloned(),
+                Key::Score => Some(MetadataValue::Float(self.record.measure as f64)),
+                // Other keys shouldn't appear (validation prevents them)
+                _ => None,
+            })
+            .collect()
+    }
+}
+
+#[async_trait]
+impl Operator<RankedGroupByInput, RankedGroupByOutput> for GroupBy {
+    type Error = RankedGroupByError;
+
+    async fn run(
+        &self,
+        input: &RankedGroupByInput,
+    ) -> Result<RankedGroupByOutput, RankedGroupByError> {
+        tracing::trace!("[RankedGroupBy] Running on {} records", input.records.len());
+
+        // Fast path: no grouping configured or no records
+        let aggregate = match &self.aggregate {
+            Some(agg) if !self.keys.is_empty() && !input.records.is_empty() => agg,
+            _ => {
+                return Ok(RankedGroupByOutput {
+                    records: input.records.clone(),
+                });
+            }
+        };
+
+        // --- Metadata hydration ---
+
+        let record_segment_reader = match Box::pin(RecordSegmentReader::from_segment(
+            &input.record_segment,
+            &input.blockfile_provider,
+        ))
+        .await
+        {
+            Ok(reader) => Some(reader),
+            Err(e) if matches!(*e, RecordSegmentReaderCreationError::UninitializedSegment) => None,
+            Err(e) => return Err((*e).into()),
+        };
+
+        let materialized_logs = materialize_logs(&record_segment_reader, input.logs.clone(), None)
+            .instrument(tracing::trace_span!(parent: Span::current(), "Materialize logs"))
+            .await?;
+
+        let offset_id_to_log = materialized_logs
+            .iter()
+            .map(|log| (log.get_offset_id(), log))
+            .collect::<HashMap<_, _>>();
+
+        let records_in_segment = input
+            .records
+            .iter()
+            .cloned()
+            .filter(|record| !offset_id_to_log.contains_key(&record.offset_id))
+            .collect::<Vec<_>>();
+
+        // Enrich records with metadata
+        let mut enriched_records = Vec::with_capacity(input.records.len());
+
+        if !records_in_segment.is_empty() {
+            let Some(reader) = &record_segment_reader else {
+                return Err(RankedGroupByError::RecordSegmentUninitialized);
+            };
+            reader
+                .load_id_to_data(records_in_segment.iter().map(|record| record.offset_id))
+                .await;
+            for record in records_in_segment {
+                let metadata = reader
+                    .get_data_for_offset_id(record.offset_id)
+                    .await?
+                    .ok_or(RankedGroupByError::PhantomRecord(record.offset_id))?
+                    .metadata
+                    .clone()
```

Comment:
Reading this diff on git, so not sure, but I think you have ownership here and could avoid this clone?